### PR TITLE
[ConstraintSystem] Prefer overloads choices that are variables over unapplied functions using the solver score.

### PR DESCRIPTION
**Pull request by hborla**
This change increases the score when the solver attempts an overload choice that is an unapplied reference to a function declaration. This will allows the solver to prune those overload choices when it has already found a solution with a property (all else equal in the score). This is already done as an ambiguity tie-breaker in solution ranking, but adding this bit to the score will prune a lot of search space within the solver.

This should fix many of the type checker performance issues that were observed with the addition of `count(where:)`, including using `array.count` in operator expressions.

Resolves: rdar://47742750

### CODE REVIEW

File: validation-test/Sema/type_checker_perf/fast/property_vs_unapplied_func.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  let date: Date
+
+  var description: String {
+    return "[" + string + "] [" + date.description + "] " + string + " [" + values.joined(separator: ", ") + "]"
+  }
+}
+
+extension Sequence {
+  public func count(
+    where predicate: (Element) throws -> Bool
```
Comment (hborla) [inline validation-test/Sema/type_checker_perf/fast/property_vs_unapplied_func.swift:21]: If anybody knows of other `count(where:)` test cases, please let me know and I'll add them to this file!

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please smoke test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Review by xedin — APPROVED** [review]

**Comment by hborla** [comment]
Source compatibility failures are unrelated (and also failing on `main`)

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please test Windows platform

**Comment by hborla** [comment]
@swift-ci please test Windows platform

**Pull request merged**